### ISSUE TITLE
Image Handler Redundancy

### ISSUE THREAD

**Issue by vsoch** [apptainer/singularity#991]
We have image handlers defined here:

https://github.com/singularityware/singularity/tree/development/libexec/handlers

and then re-done in build.exec:

https://github.com/singularityware/singularity/blob/development/libexec/cli/build.exec#L183

my concern is that there is copy pasted / redundant code, and that the "same action" (dumping docker layers) is performed in two different ways - one with permission changes, and one not. I think the files in handlers that are used by image-handler.sh should also be used by build.exec, which means:

 - deciding the appropriate base actions for each
 - moving higher level logic / specifics of the caller to the calling script
 - having the same file sourced from both places

Otherwise, over time we will have inconsistent behavior, and a code base that is harder to manage because a change would warrant updates in two+ places.

**Comment by GodloveD** [comment]
Thank you very much for pointing this out!  I wasn't aware of the image-handlers.  It would certainly be better to use them instead of duplicating a bunch of code in `build`.

Looks like some additions will need to be made to the image handlers before we can start using them in the build process.  Many of these will just be good additions to make anyway.  I'm going to put a checklist here so that I can keep track while I refactor.

- [ ] All image handlers need added capability to build into directory (probably a good idea anyway).  This is necessary so we can build into the temp staging area and then create a squashfs.
- [ ] The main case statement must be rewritten.  Right now it's depending on file extensions.  (Like Windows.  Eww.) . We now have some functions to test whether a file is a tar.*  or image or whatever.  We should them instead so that users can name their files however they like.
- [ ] Must figure out a better test for cpio format (similar to way we are looking for tar).  Magic number?
- [ ] Must write an image handler for build from local image.

Once all of these features are added to the image handlers I think it should be fine to transition build over to using them instead.

**Comment by vsoch** [comment]
woohoo!

**Comment by gmkurtzer** [comment]
I think this should go on our next release roadmap, and not for 2.4. Great stuff, thanks!

**Comment by dtrudg** [comment]
Hi @vsoch @GodloveD  - I believe this will be addressed by the Golang builder structure that @bauerm97 has been sketching out for 3.0 - do we need to keep this open to track?

**Comment by vsoch** [comment]
I agree, when the golang update associated issue is closed, I would put a note here to have this issue closed as well. If you are tracking in another issue (and all of the points here addressed) then it's safe to just close.

**Comment by GodloveD** [comment]
What @vsoch said.  😄

**Comment by decibelcooper** [comment]
Hi @vsoch @GodloveD et al, are there any updates on this?  At the moment, myself and others are having an issue with singularity/libexec/cli/build.exec, specifically that the docker image case does not correct for permissions (layers that dump read-only directories cause following layers to fail): https://github.com/singularityware/singularity/blob/9c6e8315ec5960736e283c838d824e24112d7611/libexec/cli/build.exec#L243-L261.  We would like to have a solution to the permissions issue in the near future if possible.  I am planning on submitting a PR for a simple fix for singularity/libexec/cli/build.exec, even though this issue of duplicated functionality exists.  I will reference this Issue.

**Comment by decibelcooper** [comment]
@vsoch @GodloveD nevermind!  I just realized that my problem is fixed in the development-2.x branch thanks to docker-extract: https://github.com/singularityware/singularity/blob/b1420f0e448a52a120fc364aad19d8be4dfe810e/libexec/cli/build.exec#L268.

Cheers!

**Comment by decibelcooper** [comment]
Any idea when this might be put into a release?

**Comment by dtrudg** [comment]
@decibelcooper - There is a release candidate for 2.5 out for testing: https://github.com/singularityware/singularity/releases/tag/2.5.0-rc1 - feedback would be appreciated

**Comment by dtrudg** [comment]
Closing as this is related to the deprecate 2.x codebase.
